### DISCUSSION TITLE
the threefold repetition and 50 move draw

### DISCUSSION THREAD

**Discussion by akaKarthick** [czbar/ChessForge#805] [General]
Consider adding threefold repetition and 50-move draw rules for a more accurate, rule-adherent experience. Understand there might be technical aspects, but believe the benefits for players who prefer official chess standards are worth exploring. 

![Screenshot (10)](https://github.com/czbar/ChessForge/assets/125844120/5d241ac4-f36f-4543-8ceb-4fccc7750332)
add toggle btn in this area.

btw i initially not agree with lichess dev for this feature.

https://lichess.org/forum/lichess-feedback/can-engines-understand-threefold-repetition-as-strategy#2

**Comment by czbar** [comment]
That's an interesting discussion, although I am not sure if Stockfish is to blame here.  When you look at that game on chess.com, the engine there does not claim 0.0 after the third repetition. Also, the issue does not repro for me locally on my PC.  So, perhaps this is an additional layer of logic that lichess adds and not the engine itself. I am not sure.

In any case, my view is pretty strong on this: the engine has no business claiming 0.0 after a threefold repetition. It should only care about the **rules of chess**, not about the **FIDE tournament rules** of the day. The latter do change over time. For example, the 50-rule move has been changed multiple times in my memory (e.g. at some point it was 75 moves for two knights vs a pawn).  It is the job of applications that use the engines to take care of that kind of rules.

Also, we can run into paradoxical situations (in apps like Chess Forge), when going to a certain position and turning the engine on gives, say, +5 but when evaluating the game from the beginning you would get 0.0 for that position!  Now, that would be truly confusing.

And the same applies to tablebases.  A tablebase should mark a position as winning regardless of how many moves (without a capture or a pawn move) it takes to win it. The application that uses the tablebase should apply any additional rules on top of that.

As a side note, the engine must be wary of repeating positions when one side has an advantage. Often, making a "new" move results in the same evaluation as repeating the position.   In such a case, the engine may repeat once but then has to move on rather than repeating the position forever.  This is clearly a different problem that the threefold repetition rule, though.

**Comment by akaKarthick** [comment]
Good. So we dont touch it right? 
But the lichess dev dont understand my point there.

**Comment by czbar** [comment]
As far as Chess Forge is concerned, my current preference would be not to add anything as I believe it could be more confusing than useful. In the future, perhaps we can add a Tool for threefold detection (independent of the engine) or something like that.  Let's see if a need arises.

Regarding your lichess discussion: there were 2 different people replying. Neither indicates in their lichess bio that they are developers. Do you know if they are? And, yes, there certainly seems to be a misunderstanding there.  

What I am interested in though, is whether this 0.0 evaluation indeed comes from Stockfish (rather than the web site's application layer). I know that Stockfish does keep track of the repetitions but in my tests, it never gives 0.0 due to this. I was unable to google out the answer so, when I have time, I'll check out the stockfish github discussions. (ChatGPT says Stockfish does indeed evaluate a threefold evaluation as 0.0 but you can' trust that hallucinating thing :)).

What disappoints me is that people confidently discuss these issues without apparently understanding the difference between the game rules and the tournament rules (as I already mentioned above). The way the knight moves, and the threefold repetition are referred to just as "chess rules".  They are not the same kind of rules! Btw even the threefold repetition rule has had some adjustments over time, although, admittedly, in the technical aspects (when and how to claim it) rather than in the spirit.

**Reply by akaKarthick** [reply]
I thought they are developers.

**Reply by akaKarthick** [reply]
We will add that later. Just like you said "3 fold detection".
